lib/trino-parquet/src/main/java/io/trino/parquet/reader/ChunkedInputStream.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.reader;
+
+import com.google.common.collect.Iterators;
+import io.airlift.slice.BasicSliceInput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.trino.memory.context.AggregatedMemoryContext;
+import io.trino.memory.context.LocalMemoryContext;
+import io.trino.parquet.ChunkReader;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.Iterator;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkPositionIndexes;
+import static com.google.common.io.ByteStreams.readFully;
+import static java.util.Objects.requireNonNull;
+
+public class ChunkedInputStream
+        extends InputStream
+{
+    private final LocalMemoryContext dataReaderMemoryUsage;
+    private final Iterator<BasicSliceInput> chunks;
+    private BasicSliceInput current;
+
+    public ChunkedInputStream(Collection<ChunkReader> chunks, AggregatedMemoryContext memoryContext)
```

Comment:
I think we'll need to pass in `ChunkKey` and `chunkReaders` here to ensure that we can call `free` on chunks and remove them from `chunkReaders` once we're done reading them. Otherwise chunks will stay in memory even after we're done with them until end of row group is reached, while our memory accounting is assuming that only one chunk is in memory at a time.

Reply:
If we don't remove `ChunkReader` from chunkReaders, but free it after we've done with it here, then you'll get exception about `free` getting called already from `ReferenceCountedReader` when ParquetReader tries to free up chunks on end of row group.
Could you elaborate what else do you see retained by ReferenceCountedReader that won't be accounted after this change ?

Reply:
>If we don't remove ChunkReader from chunkReaders, but free it after we've done with it here, then you'll get exception about free getting called already from ReferenceCountedReader when ParquetReader tries to free up chunks on end of row group.

looks like we should drop the `io.trino.parquet.reader.ParquetReader#freeCurrentRowGroupBuffers` and do free here after the read

> Could you elaborate what else do you see retained by ReferenceCountedReader that won't be accounted after this change ?

this only looks at the current chunk, but the actual number of retained bytes is all the `ReferenceCountedReader`s that were read but not freed so far. Am I missing something?

Reply:
> this only looks at the current chunk, but the actual number of retained bytes is all the ReferenceCountedReaders that were read but not freed so far. Am I missing something?

My suggestion was to `free` ChunkReader in ChunkInputStream when we exhaust it and move to the next one (this might need you to avoid the `Iterators.transform` and always have reference to current ChunkReader available).
This way there shouldn't be more than one `ReferenceCountedReader` active at a time and we should be accounting for it through the `dataReaderMemoryUsage.setBytes(slice.getRetainedSize())` call.

Reply:
when we merge adjacent small disk ranges w create multiple `ChunkReaders` that refer to a single `ReferenceCountedReader`. This one `ReferenceCountedReader` for a big range is retained even if we read some other columns unless all the columns that are part of this big range are read already and the free is called.

Reply:
For the case of merged ranges, my expectation is that the `slice.getRetainedSize()` will be able to catch the memory usage of the underlying large slice since all the small chunks are getting slices of the same Slice object. Let me know if you find otherwise.